### PR TITLE
upgrade: cold start nova before live migration (SOC-10761)

### PR DESCRIPTION
**Pull request by jgrassler**
In order to prevent any and all RPC API incompatibilities
through lingering nova-compute processes, this commit stops
all Nova services (both on controllers and compute nodes)
after the early nova-compute update performed at the end of
the controller upgrade. Subsequently, these services are
restarted in a staggered manner to keep them from DDoSing
RabbitMQ on larger clouds. This ensures all Nova services use
the new Nova RPC API (5.0), thus allowing the live migrations
that are part of non-disruptively upgrading the compute nodes,
to take place.

### CODE REVIEW

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb [hunk 1]
```diff
@@ -297,7 +298,24 @@
   owner "root"
   group "root"
   variables(
-    nova_controller: roles.include?("nova-controller")
+    nova_controller: roles.include?("nova-controller"),
+    remotes_present: !remote_nodes.empty?,
+    remote_node: remote_node,
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: Just a nit ... for the code clarity it might be better to rename `remote_node` to `is_remote_node`...

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb [hunk 2]
```diff
@@ -27,21 +27,41 @@ if [[ -f $UPGRADEDIR/crowbar-reload-nova-after-upgrade-ok ]] ; then
     exit 0
 fi
 
+# Check the python-nova package to make sure it is at the latest version with
+# no more recent versions in any zypper repositories (this package can be used
+# as a proxy for making sure all Nova service packages are up to date since
+# they are version locked to python-nova):
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: How could this happen? Aren't we responsible for updating the package? Repositories should be set up and the package was updated in different step, right?

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb [hunk 3]
```diff
@@ -27,21 +27,41 @@ if [[ -f $UPGRADEDIR/crowbar-reload-nova-after-upgrade-ok ]] ; then
...
-    if systemctl --quiet is-active $fullname 2>/dev/null ; then
-        systemctl restart $fullname
-    fi
+    systemctl start $fullname
+done
+
+# Apache was also stopped by crowbar-stop-nova-services.sh
+systemctl start apache2
+
+  <% if @remotes_present and @is_cluster_founder %>
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Indentation?

Comment (jsuchome) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Ah, this is part of `if @nova_controller` block...

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb [hunk 4]
```diff
@@ -27,21 +27,41 @@ if [[ -f $UPGRADEDIR/crowbar-reload-nova-after-upgrade-ok ]] ; then
...
+# Start nova-compute on remote nodes:
+for remote in $(crm resource list | awk '$1 ~ /^remote-/  {print $1}'); do
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Why not use `remote_nodes` varaible?

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb [hunk 5]
```diff
@@ -27,21 +27,41 @@ if [[ -f $UPGRADEDIR/crowbar-reload-nova-after-upgrade-ok ]] ; then
...
+  crm resource start $remote
 done
-# Remove and unmanage openstack-nova-consoleauth
-systemctl disable openstack-nova-consoleauth
-systemctl stop openstack-nova-consoleauth
-systemctl kill openstack-nova-consoleauth
-rpm -e openstack-nova-consoleauth
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: I assume this can be deleted as the service was already removed in SOC8, right?

Comment (jsuchome) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Ah, no, you've moved it to the other script.

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+systemctl stop openstack-nova-compute
+  <% end %>
+<% end %>
+
+# This check can run on both remote and non-remote nodes: the `crm resource
+# stop` for the remote resources will have been run before this script runs on
+# the compute node it was intended for. This check needs to fail on the compute
+# node itself if it is going to fail. That way it will pinpoint the problematic
+# node.
+if ps auxw | grep ^nova | grep python; then
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: Hm, are we sure systemctl stop already killed every such process? Shouldn't there be some short wait loop?

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb [hunk 7]
```diff
@@ -297,7 +298,24 @@
   owner "root"
   group "root"
   variables(
-    nova_controller: roles.include?("nova-controller")
+    nova_controller: roles.include?("nova-controller"),
+    remotes_present: !remote_nodes.empty?,
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: What about `remote_nodes.any?`

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb (hunk 6 above)
Comment (jsuchome) [reply chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: Yep, my test just failed with `'Error: there are still nova processes running. Stop these processes before continuing the upgrade.'`
I logged in, checked manually and no nova process is there. I guess we need to give it some time..

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb (hunk 2 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: We are, but the `zypper --non-interactive up openstack-nova-compute` we run for the early `nova-compute` upgrade we require for RPC API compatibility is not trustworthy. For instance, it exited `0` when we'd forgotten to set `solver.allowVendorChange` earlier, thus causing the installation to fail with a prompt asking about vendor change. I don't know how `zypper` behaves with other problems, such as a NIC flap preventing access to the package repository, but I'd rather play it safe and check up on it.

Comment (jsuchome) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: OK, then. It does not hurt to have another check I guess

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb (hunk 4 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Because I'd be building the resource name from the host name in there and a prepended `remote-`. I'm not comfortable assuming that this is always the correct resource name. For instance, what happens if there's a FQDN in the list and a short host name in pacemaker (or vice versa)? I can rely on whatever matches that regex in the pacemaker resource list being a remote resource since pacemaker is the canonical source of resource names.

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb (hunk 1 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: Done.

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb (hunk 6 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: Ok, I added a retry loop now.

File: chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb (hunk 7 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/recipes/prepare-upgrade-scripts.rb]: Done.

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+<% end %>
+
+# This check can run on both remote and non-remote nodes: the `crm resource
+# stop` for the remote resources will have been run before this script runs on
+# the compute node it was intended for. This check needs to fail on the compute
+# node itself if it is going to fail. That way it will pinpoint the problematic
+# node.
+
+for wait in 10 20 40 0; do
+  if ps -U nova u | grep nova-compute; then
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: It's not just for compute services, right?

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb (hunk 4 above)
Comment (jsuchome) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: OK

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+<% end %>
+
+# This check can run on both remote and non-remote nodes: the `crm resource
+# stop` for the remote resources will have been run before this script runs on
+# the compute node it was intended for. This check needs to fail on the compute
+# node itself if it is going to fail. That way it will pinpoint the problematic
+# node.
+
+for wait in 10 20 40 0; do
+  if [ $(ps -U nova u | wc -l) -gt 1 ]; then
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: This is actually still wrong:

```
# ps -U nova u 
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
nova     28717  0.0  0.1  36912  4656 ?        Ss   14:24   0:00 /usr/lib/systemd/systemd --user
nova     28720  0.0  0.0 166136  1980 ?        S    14:24   0:00 (sd-pam)
```

Comment (jgrassler) [reply chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: Sigh. This is a veritable Rubik's Cube :-)

Modified again...and that looks like it does the trick on both controller...

```
root@d00-25-90-e5-04-54:~ # ps -U nova u | grep -e apache -e python
nova      3811  0.0  0.5 307760 95552 ?        Ss   Oct30   0:17 /usr/bin/python /usr/bin/nova-novncproxy
nova      8240  0.0  0.6 593152 111972 ?       Sl   Oct29   0:28 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8241  0.0  0.6 593144 110904 ?       Sl   Oct29   0:27 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8242  0.0  0.6 658704 111116 ?       Sl   Oct29   0:28 (wsgi:nova-placement-ap -DSYSCONFIG -C PidFile /var/run/httpd.pid -C Include /etc/apache2/sysconfig.d//loadmodule.conf -C Include /etc/apache2/sysconfig.d//global.conf -f /etc/apache2/httpd.conf -c Include /etc/apache2/sysconfig.d//include.conf -DSYSTEMD -DFOREGROUND -k start
nova      8291  1.1  0.6 274660 100344 ?       Ss   Oct29  27:12 /usr/bin/python /usr/bin/nova-conductor
nova      8517  0.0  0.6 288568 113976 ?       Ss   Oct29   2:02 /usr/bin/python /usr/bin/nova-scheduler
nova      8624  1.1  0.7 311244 130692 ?       Ss   Oct29  27:26 /usr/bin/python /usr/bin/nova-api
nova      8807  0.0  0.6 284080 109700 ?       Ss   Oct29   1:41 /usr/bin/python /usr/bin/nova-consoleauth
nova     24523  0.0  0.8 323888 137768 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24527  0.0  0.8 322960 136960 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24528  0.0  0.8 323924 137784 ?       S    Oct30   0:01 /usr/bin/python /usr/bin/nova-api
nova     24529  0.0  0.8 324360 137868 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24530  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24531  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24532  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24533  0.0  0.8 318868 131488 ?       S    Oct30   0:00 /usr/bin/python /usr/bin/nova-api
nova     24577  0.2  0.7 297400 117364 ?       S    Oct30   2:07 /usr/bin/python /usr/bin/nova-conductor
nova     24578  0.2  0.7 297476 117592 ?       S    Oct30   2:08 /usr/bin/python /usr/bin/nova-conductor
nova     24585  0.2  0.7 296896 116796 ?       S    Oct30   2:05 /usr/bin/python /usr/bin/nova-conductor
nova     24586  0.2  0.7 297032 116952 ?       S    Oct30   2:09 /usr/bin/python /usr/bin/nova-conductor
root@d00-25-90-e5-04-54:~ #
```

...and compute node:

```
root@d00-25-90-e5-0c-fc:~ # ps -U nova u | grep -e apache -e python
nova     13415  0.2  0.7 1961284 130156 ?      Ssl  Oct30   2:30 /usr/bin/python /usr/bin/nova-compute
nova     13667  0.2  0.7 1961524 129976 ?      Ssl  Oct30   2:28 /usr/bin/python /usr/bin/nova-compute
root@d00-25-90-e5-0c-fc:~ #
```

File: chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+# Remove and unmanage openstack-nova-consoleauth
+systemctl disable openstack-nova-consoleauth
+systemctl stop openstack-nova-consoleauth
+systemctl kill openstack-nova-consoleauth
+rpm -e openstack-nova-consoleauth
+
+<% else %>
+  # Only stop on non-remote compute nodes - service on remote nodes will be stopped by
+  # the controller side code above.
```
Comment (jsuchome) [inline chef/cookbooks/crowbar/templates/default/crowbar-stop-nova-services.sh.erb]: Could you move the comment into the `unless` block?

This way it is rendered with the comment only on remote nodes and is not clear what does the comment refer to. So it's better to not show it in the script (while it is perfectly good in template)

**Comment by jsuchome** [comment]
will need a rebase after https://github.com/crowbar/crowbar-core/pull/1950 is merged

**Comment by jsuchome** [comment]
Please rebase

File: crowbar_framework/app/models/api/upgrade.rb [hunk 11]
```diff
@@ -817,6 +817,8 @@ def do_controllers_substep(substep)
           upgrade_controller_clusters
           upgrade_non_compute_nodes
           prepare_all_compute_nodes
+          # Stop nova services on all nodes to prevent any and all RPC API trouble
+          stop_nova_services
           reload_nova_services
```
Comment (skazi0) [inline crowbar_framework/app/models/api/upgrade.rb]: this `reload_nova_services` call doesn't make much sense here now, does it?

Comment (jsuchome) [reply crowbar_framework/app/models/api/upgrade.rb]: It starts the services. So it still makes sense, though the naming might be bit confusing

Comment (jgrassler) [reply crowbar_framework/app/models/api/upgrade.rb]: I didn't change the name on purpose - didn't want to risk breaking something due to a dangling reference to the old name somewhere...

**Comment by jgrassler** [comment]
> Please rebase

Done.

File: crowbar_framework/app/models/api/upgrade.rb (hunk 11 above)
Comment (skazi0) [reply crowbar_framework/app/models/api/upgrade.rb]: ah, so the idea here is that we stop everything and then start it all again without anything in between, correct? if so then I missed the point before.

Comment (jgrassler) [reply crowbar_framework/app/models/api/upgrade.rb]: Yes.

**Comment by jgrassler** [comment]
This needed another rebase now due to landing (the merge conflict only affected the cleanup line for the "run-nova-online-migrations" script).

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb [hunk 12]
```diff
@@ -27,21 +27,41 @@ if [[ -f $UPGRADEDIR/crowbar-reload-nova-after-upgrade-ok ]] ; then
...
+  exit 1
+fi
+
 <% if @nova_controller %>
 for service in conductor scheduler novncproxy serialproxy api; do
     fullname="openstack-nova-$service"
-    if systemctl --quiet is-active $fullname 2>/dev/null ; then
-        systemctl restart $fullname
-    fi
+    systemctl start $fullname
```
Comment (skazi0) [inline chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: why did you remove the `is-active` check? do we always have all of these services installed? if something is missing then you should not try to start it.

Comment (skazi0) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: `if systemctl cat $fullname &>/dev/null; then ...` seems to do the trick.

File: crowbar_framework/lib/crowbar/upgrade_timeouts.rb [hunk 13]
```diff
@@ -38,6 +38,7 @@ def values
         delete_nova_services: @timeouts_config[:delete_nova_services] || 300,
         wait_until_compute_started: @timeouts_config[:wait_until_compute_started] || 60,
         reload_nova_services: @timeouts_config[:reload_nova_services] || 120,
+        stop_nova_services: @timeouts_config[:stop_nova_services] || 120,
```
Comment (skazi0) [inline crowbar_framework/lib/crowbar/upgrade_timeouts.rb]: Add this to `configs/upgrade_timeouts.yml` as well. This file is packaged as an example for the users.

File: chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb (hunk 12 above)
Comment (jgrassler) [reply chef/cookbooks/crowbar/templates/default/crowbar-reload-nova-after-upgrade.sh.erb]: Done.

File: crowbar_framework/lib/crowbar/upgrade_timeouts.rb (hunk 13 above)
Comment (jgrassler) [reply crowbar_framework/lib/crowbar/upgrade_timeouts.rb]: Done.
